src/pages/workspaces/workspace/WorkspaceMenu.js

```diff
@@ -0,0 +1,90 @@
+import { Fragment } from 'react'
+import { h } from 'react-hyperscript-helpers'
+import { Clickable } from 'src/components/common'
+import { icon } from 'src/components/icons'
+import { makeMenuIcon, MenuButton, MenuTrigger } from 'src/components/PopupTrigger'
+import { useWorkspaceDetails } from 'src/components/workspace-utils'
+import * as Utils from 'src/libs/utils'
+
+
+// In `workspaceInfo`, specify either `name and namespace` to fetch the Workspace details,
+// or `canShare, isAzureWorkspace, isLocked, and isOwner` to use previously fetched details.
+const WorkspaceMenu = ({
+  iconSize, popupLocation,
+  callbacks: { onClone, onShare, onLock, onDelete },
+  workspaceInfo: { name, namespace, canShare, isAzureWorkspace, isLocked, isOwner }
+}) => {
+  const navIconProps = {
+    style: { opacity: 0.65, marginRight: '1rem' },
+    hover: { opacity: 1 }, focus: 'hover'
+  }
+
+  const menuContent = !!namespace ?
+    h(DynamicWorkspaceMenuContent, { namespace, name, onShare, onClone, onDelete, onLock }) :
+    h(WorkspaceMenuContent, { canShare, isAzureWorkspace, isLocked, isOwner, onClone, onShare, onLock, onDelete })
+
+  return h(MenuTrigger, {
+    side: popupLocation,
+    closeOnClick: true,
+    content: menuContent
+  }, [
+    h(Clickable, {
+      'aria-label': !!name ? `Action Menu for Workspace: ${name}` : 'Workspace Action Menu',
+      'aria-haspopup': 'menu',
+      ...navIconProps
+    }, [icon('cardMenuIcon', { size: iconSize })])
+  ])
+}
+
+const DynamicWorkspaceMenuContent = ({ namespace, name, onClone, onShare, onDelete, onLock }) => {
+  const { workspace } = useWorkspaceDetails({ namespace, name }, ['accessLevel', 'azureContext', 'canShare', 'workspace.isLocked'])
+  const canShare = workspace?.canShare
+  const isOwner = workspace && Utils.isOwner(workspace.accessLevel)
+  const isLocked = workspace?.workspace.isLocked
+  const isAzureWorkspace = !!workspace?.azureContext
+  return WorkspaceMenuContent({
+    canShare, isAzureWorkspace, isLocked, isOwner, onClone, onShare, onLock, onDelete, workspaceLoaded: !!workspace
```

Comment:
Would it be preferable to wait until the workspace is loaded to render the menu content? The initial disabled states can be confusing if the workspace details call is slow.

Reply:
I agree that it isn't a somewhat confusing user experience, but I think changing it is beyond the scope of this ticket (given that it's already this way in production). I will create a Workspaces ticket to consult with UX and look at options. I think it would be odd to show a page-wide spinner when you open the menu, and really it would be best to have this state upfront so we could display the status of the workspaces in the list table itself (currently you have to open the menu to see if a workspace is locked, for example, and we want to be able to show the cloud context for the workspace in the list). Therefore I think we would want to investigate the feasibility of adding cloud context, locked, and canShare to the workspaces list API (accessLevel is already in that API).

Reply:
Sounds good. Agreed about the page-wide spinner, I was thinking something like a small spinner in the popup. But returning the data from the workspaces list API does sound like the ideal solution.

Reply:
I created https://broadworkbench.atlassian.net/browse/WOR-316. We already have a ticket in an upcoming sprint to add the cloud context info to workspaces list, so hopefully we can tackle `locked` and `canShare` at the same time.

Reply:
@cahrens `isLocked` is already available from the listWorkspaces API, but the UI excludes it from the API response by only requesting a limited number of workspace fields. The history behind this is that we wanted to reduce the size of the JSON response returned, since it can be pretty large. However, I think this is a good reason for the UI to start requesting this field. It should just be a matter of adding `workspace.isLocked` to the list of fields [here](https://github.com/DataBiosphere/terra-ui/blob/dev/src/components/workspace-utils.js#L32)

Reply:
Great! That means that `canShare` and cloud context are the only things we would need to add to the list API.